File: langs/java/backend/src/main/java/dev/argon/nobleidl/compiler/PackageMappingScannerProcessor.java
package dev.argon.nobleidl.compiler;

import dev.argon.esexpr.KeywordMapping;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.*;
import java.util.*;

@SupportedAnnotationTypes(PackageMappingScannerProcessor.packageAnnName)
final class PackageMappingScannerProcessor extends AbstractProcessor {

	final static String packageAnnName = "dev.argon.nobleidl.runtime.NobleIDLPackage";

	private final Map<String, String> packageMapping = new HashMap<>();

	public PackageMapping getPackageMapping() {
		return new PackageMapping(new KeywordMapping<>(new HashMap<>(packageMapping)));
	}


	@Override
	public SourceVersion getSupportedSourceVersion() {
		return SourceVersion.latestSupported();
	}

	@Override
	public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
		for(TypeElement annotation : annotations) {
			if(!annotation.getQualifiedName().toString().equals(PackageMappingScannerProcessor.packageAnnName)) {
				continue;
			}

			Set<? extends Element> annotatedElements = roundEnv.getElementsAnnotatedWith(annotation);

			for(var element : annotatedElements) {
				if(!(element instanceof PackageElement packageElement)) {
					continue;
				}

				var mappingAnn = getAnnotation(element.getAnnotationMirrors(), PackageMappingScannerProcessor.packageAnnName).orElse(null);
				if(mappingAnn == null) {
					continue;
				}

				var idlPackage = getAnnotationArgument(mappingAnn, "value").orElse(null);
				if(idlPackage == null) {
					continue;
				}

				if(!(idlPackage.getValue() instanceof String idlPackageValue)) {
					continue;
				}

				packageMapping.put(idlPackageValue, packageElement.getQualifiedName().toString());
			}

		}

		return true;
	}


	private static Optional<? extends AnnotationMirror> getAnnotation(List<? extends AnnotationMirror> annotations, String name) {
		return annotations.stream()
			.filter(ann -> ((TypeElement)ann.getAnnotationType().asElement()).getQualifiedName().toString().equals(name))
			.findFirst();
	}

	private static Optional<AnnotationValue> getAnnotationArgument(AnnotationMirror ann, String name) {
		return ann.getElementValues()
			.entrySet()
			.stream()
			.filter(entry -> entry.getKey().getSimpleName().toString().equals(name))
			.findFirst()
			.map(Map.Entry::getValue);
	}
}

File: langs/scala/compiler/src/main/scala/nobleidl/compiler/PackageMappingScannerProcessor.scala
package nobleidl.compiler

import nobleidl.compiler.backend.PackageMapping

import esexpr.Dictionary
import javax.annotation.processing.{AbstractProcessor, ProcessingEnvironment, RoundEnvironment, SupportedAnnotationTypes}
import javax.lang.model.SourceVersion
import javax.lang.model.element.*
import scala.collection.mutable
import scala.jdk.CollectionConverters.*

@SupportedAnnotationTypes(Array(
  "nobleidl.core.NobleIDLScalaPackage",
  "dev.argon.nobleidl.runtime.NobleIDLPackage",
  "nobleidl.sjs.core.NobleIDLScalaJSPackage",
  "nobleidl.sjs.core.NobleIDLScalaJSImport",
))
final class PackageMappingScannerProcessor extends AbstractProcessor {
  private final val scalaPackageMapping = mutable.Map[String, String]()
  private final val javaPackageMapping = mutable.Map[String, String]()
  private final val scalaJSPackageMapping = mutable.Map[String, String]()
  private final val scalaJSImportMapping = mutable.Map[String, String]()

  def getLibraryResults: LibraryAnalyzer.LibraryResults =
    LibraryAnalyzer.LibraryResults(
      scalaPackageMapping = scalaPackageMapping.toMap,
      javaPackageMapping = javaPackageMapping.toMap,
      scalaJSPackageMapping = scalaJSPackageMapping.toMap,
      scalaJSImportMapping = scalaJSImportMapping.toMap,
      sourceFiles = Seq.empty,
    )

  override def getSupportedSourceVersion: SourceVersion = SourceVersion.latestSupported

  override def process(annotations: java.util.Set[? <: TypeElement], roundEnv: RoundEnvironment): Boolean = {
    def processMappingFunc(annName: String)(addMapping: (String, String) => Unit): Unit =
      for annotation <- annotations.asScala.view.filter(_.getQualifiedName.toString == annName) do
        val annotatedElements = roundEnv.getElementsAnnotatedWith(annotation)
        
        for
          case element: PackageElement <- annotatedElements.asScala
          mappingAnn <- getAnnotation(element.getAnnotationMirrors.nn.asScala, annName)
          idlPackageValue <- getAnnotationArgument(mappingAnn, "value")
          case idlPackage: String <- Seq(idlPackageValue.getValue)
        do addMapping(idlPackage, element.getQualifiedName.toString)
      end for
      
    def processMapping(annName: String, mapping: mutable.Map[String, String]): Unit =
      processMappingFunc(annName)(mapping.update)

    processMapping("nobleidl.core.NobleIDLScalaPackage", scalaPackageMapping)
    processMapping("dev.argon.nobleidl.runtime.NobleIDLPackage", javaPackageMapping)
    processMapping("nobleidl.sjs.core.NobleIDLScalaJSPackage", scalaJSPackageMapping)

    processMappingFunc("nobleidl.sjs.core.NobleIDLScalaJSImport") { (importPath, pkg) =>
      scalaJSPackageMapping.find(_._2 == pkg).foreach { (idlPackage, _) =>
        scalaJSImportMapping(idlPackage) = importPath
      }
    }

    true
  }

  private def getAnnotation(annotations: Iterable[AnnotationMirror], name: String): Option[AnnotationMirror] =
    annotations.find { ann => ann.getAnnotationType.asElement.asInstanceOf[TypeElement].getQualifiedName.toString == name }

  private def getAnnotationArgument(ann: AnnotationMirror, name: String): Option[AnnotationValue] =
    ann.getElementValues.nn
      .asScala
      .find { (k, _) => k.getSimpleName.toString == name }
      .map { (_, v) => v }
}